torch/csrc/autograd/profiler.cpp

```diff
@@ -25,6 +25,23 @@ void enableProfiler(ProfilerState new_state) {
   }
   state = new_state;
   mark("__start_profile");
+
+#ifdef WITH_CUDA
+  if(state == ProfilerState::CUDA) {
+    // cuda events must be on the same device, so we need a start event recorded
+    // for each gpu
+    int old_device;
+    // this is not exactly AutoGPU because we need to iterate through each GPU
+    TORCH_CUDA_CHECK(cudaGetDevice(&old_device));
```

Comment:
Use `AutoGPU` for this.

Reply:
I don't think it works. If I don't specify a _new_ device when I want to save the old one, it will just no-op. If I use AutoGPU everywhere, then I make twice as many setDevice calls.

```
  inline void setDevice(int device) {
#ifdef WITH_CUDA
    if (device == -1) {
      return;
    }
```

Reply:
Sorry, I didn’t understand that. I can’t see how auto GPU will increase the number of set device calls